superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx

```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
+      stub = sinon.stub(shortid, 'generate').returns('abcd');
+    });
+    afterEach(() => {
+      stub.restore();
+    });
+
+    it('should generate data source name from query', () => {
+      const sampleQuery = queries[0];
+      const name = wrapper.instance().datasourceName();
+      expect(name).to.equal(`${sampleQuery.user}-${sampleQuery.db}-${sampleQuery.tab}-abcd`);
+    });
+    it('should generate data source name with empty query', () => {
+      wrapper.setProps({ query: {} });
+      const name = wrapper.instance().datasourceName();
+      expect(name).to.equal('undefined-abcd');
+    });
+  });
+
+  describe('mergedColumns', () => {
+    const wrapper = getVisualizeModalWrapper();
+    const oldColumns = {
+      ds: 1,
+      gender: 2,
+    };
+
+    it('should merge by column name', () => {
+      wrapper.setState({ columns: {} });
+      const mc = wrapper.instance().mergedColumns();
+      expect(mc).to.deep.equal(mockColumns);
+    });
+    it('should not override current state', () => {
+      wrapper.setState({ columns: oldColumns });
+
+      const mc = wrapper.instance().mergedColumns();
+      expect(mc.ds).to.equal(oldColumns.ds);
+      expect(mc.gender).to.equal(oldColumns.gender);
+    });
+  });
+
+  describe('validate', () => {
+    const wrapper = getVisualizeModalWrapper();
+    let columnsStub;
+    beforeEach(() => {
+      columnsStub = sinon.stub(wrapper.instance(), 'mergedColumns');
+    });
+    afterEach(() => {
+      columnsStub.restore();
+    });
+
+    it('should validate column name', () => {
+      columnsStub.returns(mockColumns);
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(0);
+      wrapper.instance().mergedColumns.restore();
+    });
+    it('should hint invalid column name', () => {
+      columnsStub.returns({
+        '&': 1,
+      });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
+      wrapper.instance().mergedColumns.restore();
     });
+    it('should hint empty chartType', () => {
+      columnsStub.returns(mockColumns);
+      wrapper.setState({ chartType: null });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
+    });
+    it('should check time series', () => {
+      columnsStub.returns(mockColumns);
+      wrapper.setState({ chartType: mockChartTypeTB });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(0);
+
+      // no is_date columns
+      columnsStub.returns({
+        ds: {
+          is_date: false,
+          is_dim: false,
+          name: 'ds',
+          type: 'STRING',
+        },
+        gender: {
+          is_date: false,
+          is_dim: true,
+          name: 'gender',
+          type: 'STRING',
+        },
+      });
+      wrapper.setState({ chartType: mockChartTypeTB });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
```

Comment:
let's check this hint string as well like you did here: https://github.com/airbnb/superset/pull/2811/files#diff-5fd795ed9490b587cec77d05effde1d1R169